### ISSUE TITLE
False positive null pointer exception

### ISSUE THREAD

**Issue by Simbiat** [kalessil/phpinspectionsea#1946]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) |
| Language level | e.g. PHP 8.3                        |

Current behaviour
--
Have a DOMNode element and code like this
```php
$node?->parentNode->removeChild($node)
```
Get null pointer exception warning
![image](https://github.com/kalessil/phpinspectionsea/assets/6022665/1d15936c-1fc6-45fa-82fc-35025aa9391b)

Expected behaviour
--
No warning, since there is is a null safe operator `?->` used for the same object.

Environment details
---

```
PhpStorm 2024.1.1
Build #PS-241.15989.102, built on April 23, 2024
Licensed to simbiat.ru / Dmitry Kustov
Subscription is active until March 31, 2025.
For non-commercial open source development only.
Runtime version: 17.0.10+1-b1207.14 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11.0
GC: G1 Young Generation, G1 Old Generation
Memory: 5120M
Cores: 16
Registry:
  debugger.new.tool.window.layout=true
  run.processes.with.pty=TRUE
  ide.experimental.ui=true
  terminal.new.ui=true
Non-Bundled Plugins:
  com.intellij.ml.llm (241.15989.156)
  com.kalessil.phpStorm.phpInspectionsEA (5.0.0.0)
  org.intellij.qodana (241.15989.121)
```

**Comment by jtheuerkauf** [comment]
Similar example in a `match` statement:

```php
function foo (?SomeEnum $bar) {
    match ($bar) {
        null => 'invalid',
        SomeEnum::SomeCase => $bar->value,
    };
}
```
Accessing `$bar->value` gets an inspection flag. However, the `null` case is evaluated first, making it impossible for `$bar` to be `null` in the second case.

![image](https://github.com/kalessil/phpinspectionsea/assets/105436210/23eee692-e6b5-4981-9cc0-bdaeabd2ebd3)

```
PhpStorm 2024.2 EAP
Build #PS-242.10180.24, built on May 14, 2024
Runtime version: 21.0.3+13-b446.1 amd64 (JCEF 122.1.9)
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11.0
GC: G1 Young Generation, G1 Concurrent GC, G1 Old Generation
Memory: 4096M
Cores: 12
Registry:
  debugger.new.tool.window.layout=true
  run.processes.with.pty=TRUE
  ide.experimental.ui=true
```

**Comment by janhanacek1** [comment]
I have similiar problem in ternary operator usage.

![image](https://github.com/kalessil/phpinspectionsea/assets/73661684/bafcf351-36c7-4655-8d17-d24fe68bc186)

**Comment by janhanacek1** [comment]
I have similiar problem in ternary operator usage...

![image](https://github.com/kalessil/phpinspectionsea/assets/73661684/122cf5d0-5475-40dc-b85c-cfd2d360f364)

**Comment by chriscpty** [comment]
i think this can be generalised as a false positive whenever the check for null is done indirectly (so not something along the lines of `$foo !== null` or `empty($foo)`).
If the null/empty check instead checks for one of `$foo`s attributes (through `?->` or similar), the inspection doesn't pick up on it.

I think @jtheuerkauf 's `match` error is a separate issue?

**Comment by Aeliot-Tm** [comment]
There same problem with [Doctrine/Criteria](https://github.com/doctrine/collections)
<img width="960" height="408" alt="Image" src="https://github.com/user-attachments/assets/0e6c679c-3967-4680-a660-78a51ebcf32c" />

Even called static method 'expr' contains [builder inside](https://github.com/doctrine/collections/blob/2.4.x/src/Criteria.php#L54)

<img width="960" height="408" alt="Image" src="https://github.com/user-attachments/assets/29022be4-a64d-47ba-80dd-b4c8b135db00" />
